hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+        case _ => ((nme: String) => VarSymbol(Tree.Ident(nme)))
+      
+      val capturesSymbols = includedCaptures.map: sym =>
+        (sym, createSym(sym.nme + "$capture"))
+
+      val localsSymbols = includedLocals.map: sym =>
+        (sym, createSym(sym.nme))
+
+      val isymSymbols = clsCaptures.map: sym =>
+        (sym, createSym(sym.nme + "$instance"))
+
+      val bmsSymbols = reqdBms.map: sym =>
+        (sym, createSym(sym.nme + "$member"))
+
+      val extraParamsCaptures = capturesSymbols.map: // parameter list
+        case (d, sym) => Param(FldFlags.empty, sym, None)
+      val newCapturePaths = capturesSymbols.map: // mapping from sym to param symbol
+        case (d, sym) => d -> sym.asPath
+      .toMap
+
+      val extraParamsLocals = localsSymbols.map: // parameter list
+        case (d, sym) => Param(FldFlags.empty, sym, None)
+      val newLocalsPaths = localsSymbols.map: // mapping from sym to param symbol
+        case (d, sym) => d -> sym
+      .toMap
+
+      val extraParamsIsyms = isymSymbols.map: // parameter list
+        case (d, sym) => Param(FldFlags.empty, sym, None)
+      val newIsymPaths = isymSymbols.map: // mapping from sym to param symbol
+        case (d, sym) => d -> sym
+      .toMap
+
+      val extraParamsBms = bmsSymbols.map: // parameter list
+        case (d, sym) => Param(FldFlags.empty, sym, None)
+      val newBmsPaths = bmsSymbols.map: // mapping from sym to param symbol
+        case (d, sym) => d -> sym
+      .toMap
+
+      val extraParams = extraParamsBms ++ extraParamsIsyms ++ extraParamsLocals ++ extraParamsCaptures
+
+      val newCtx = ctx
+        .replCapturePaths(newCapturePaths)
+        .replLocalPaths(newLocalsPaths)
+        .addIsymPaths(newIsymPaths)
+        .replIgnoredBmsPaths(newBmsPaths)
+
+      d match
+        case f: FunDefn =>
+          // create second param list with different symbols
+          val extraParamsCpy = extraParams.map(p => p.copy(sym = VarSymbol(p.sym.id)))
+
+          val headPlistCopy = f.params.headOption match
+            case None => PlainParamList(Nil)
+            case Some(value) => ParamList(value.flags, value.params.map(p => p.copy(sym = VarSymbol(p.sym.id))), value.restParam)
+          
+          val flatPlist = f.params match
+            case head :: next => ParamList(head.flags, extraParams ++ head.params, head.restParam) :: next
+            case Nil => PlainParamList(extraParams) :: Nil
+
+          val newDef = FunDefn(
+            base.owner, f.sym, PlainParamList(extraParams) :: f.params, f.body
+          )
+          val Lifted(lifted, extras) = liftDefnsInFn(newDef, newCtx)
+
+          val args1 = extraParamsCpy.map(p => p.sym.asPath.asArg)
+          val args2 = headPlistCopy.params.map(p => p.sym.asPath.asArg)
+
+          val bdy = blockBuilder
+            .ret(Call(singleCallBms.asPath, args1 ++ args2)(true, false)) // TODO: restParams not considered
```

Comment:
Can you add a test where this TODO causes a failure? And please add an item to https://github.com/hkust-taco/mlscript/issues/285